### DISCUSSION TITLE
Digest or Named reference signature files from images with Connaisseur

### DISCUSSION THREAD

**Discussion by allanedwardslab651** [sse-secure-systems/connaisseur#1364] [Q&A]
How does Connaisseur find the signature files that are related to our image artifacts?  Also, is there any way to configure Connaisseur to look into another repo folder for a sig file from where the image is stored?

**Comment by Starkteetje** [comment]
Hi @allanedwardslab651 
it would be helpful if you were a bit more specific. Different signature schemes work differently. However, from the other question you posed, I'd presume you're using Cosign. In that case, Cosign stores signature data alongside the image, [c.f. their documentation](https://github.com/sigstore/cosign#sign-a-container-and-store-the-signature-in-the-registry).

In case you're using Notary, there's a dedicated Notary server needed. However, docker.io does have one by default, so if you're using images from its library, you can use them as Docker's public key and notary service is pre-configured in our default configuration [here](https://github.com/sse-secure-systems/connaisseur/blob/master/helm/values.yaml#L138-L149). See further documentation [here](https://github.com/notaryproject/notary/blob/master/docs/getting_started.md).

**Reply by allanedwardslab651** [reply]
My apologies for not being more specific.  We are utilizing cosign.   Some of our developers are storing their signature files in a .sig repo away from the where the image is stored.  We are trying to figure out if there is a way to store our signature files in a separate repo from the repo we store our images in.  We did not find anything in your documentation that specified how Connaisseur located signature files from images during the verification process.  Do you have any additional information on how Connaisseur locates the signature file from the image provided.

**Reply by Starkteetje** [reply]
No worries and my apologies for my previous answer being brusque.

Since Connaisseur is (currently) written in Python and Cosign is written in Go, we're not delving into Cosign's internals, but rather call it's binary, which is part of the Connaisseur image. Currently, we have a standardized way of calling Cosign ([here](https://github.com/sse-secure-systems/connaisseur/blob/master/connaisseur/validators/cosign/cosign_validator.py#L298)), which basically boils down to `cosign verify --key ${YOUR_TRUST_ROOT} ${YOUR_IMAGE}`. That makes use of Cosign's design that allows signatures to be stored alongside the image in the registry.
Given you have signature artifacts outside the registry, I don't think this can be made to work with our current way of handling Cosign. Generally, I could see us adding an option for something like `cosign verify-blob --signature ${SIGNATURE} ${YOUR_IMAGE}` though, which would accomodate your use case. However, I'm not sure whether there's a suitable, generic mapping that would allow us to discover the signature at runtime (which I guess is the root of your question).

The basic problem is how to get the signature from just the image reference. When receiving the admission webhook, Connaisseur effectively obtains the referenced image from the k8s manifest, which doesn't naturally transport any trust data such as signatures (and I don't really see that changing). The usual Cosign flow does it by storing the signature within the registry, while Notary has a dedicated interface, whose host can be configured, so we know how to obtain trust data from there. With "arbitrary" signature files (which I guess the signatures are, unless they're stored based on some standard?), my naiive approach would be to allow configuration of a signature_getter with a remote host and some mapping from image to request path for a http request, but that screams complexity. Maybe you have better ideas here on how to solve this or how to simplify the problem?

Otherwise, your case of storing signatures in a repo not affiliated with the registry sounds somewhat akin to the Notary setup, so depending on your needs, the setup of a Notary server for your devs might solve your security requirements?

**Reply by Starkteetje** [reply]
Actually, @xopham knows more about the Cosign ecosystem than I do and dug up https://docs.sigstore.dev/system_config/registry_support/#specifying-registry. Connaisseur has an extra [env field](https://github.com/sse-secure-systems/connaisseur/blob/master/helm/values.yaml#L58) in its config, so passing `COSIGN_REPOSITORY` should work and impact the execution of the binary, so if the repository you're hosting the signatures in has the correct structure, this could be a solution that works with both your approach and as-is-Connaisseur.

Could that approach work for you? I'd be interested to hear, whether it does or if not, what blocks it from working

**Reply by allanedwardslab651** [reply]
Yes, we were using COSIGN_REPOSITORY and specifying a different repository for our sig files.  This was causing Connaisseur to not work properly.  It expects the sig file in the same repo as the image we found out through testing.

**Reply by Starkteetje** [reply]
Hmm, that's weird. I just tried it and for me it did work to have an image at https://hub.docker.com/layers/teetje/cosigntest/1.2.3/images/sha256-90659bf80b44ce6be8234e6ff90a1ac34acbeb826903b02cfa0da11c82cbc042?context=repo validated by a signature in [another repo](https://hub.docker.com/layers/teetje/cosigntestredirect/sha256-90659bf80b44ce6be8234e6ff90a1ac34acbeb826903b02cfa0da11c82cbc042.sig/images/sha256-f5662dddd730f2534ae64490991c5b66d1999c8192b20db0874fef890213da7f?context=repo), so in general this seems to work. Maybe something else is not working due to it being a totally different registry instead if just another repository in your setup.

We'll most likely release a new version of Connaisseur today, which will include debug logging for Cosign validators when Connaisseur is run with debug log level. Once that's released, it would be nice if you could test in debug mode and share (parts of) the log output and your `values.yaml` config, so that I can better understand what is going wrong in your instance

**Reply by xopham** [reply]
@allanedwardslab651 have you tested the solution to set the repository via connaisseurs environment variable field as proposed by @Starkteetje above?
@Starkteetje one limitation I see is the limitation to only one signature repo and all others will fail. Specifically, if you have multiple cosign validators for external and internal images. we might have to make the signature repository configurable nontheless

**Reply by Starkteetje** [reply]
A day later than we'd hoped, but we've just released [v3.3.0](https://github.com/sse-secure-systems/connaisseur/releases/tag/v3.3.0), which does enabled verbose logging of Cosign, if Connaisseur's log level is set to debug. With that, we should get an answer as to why setting `COSIGN_REPOSITORY` worked for me, but not in your setup. It'd be nice if you could share these logs with us
